Extending ORCA blockplotting (w/ tests)

lbdreyer opened the pull request:
This adds some changes to the testing of the 2d blockplotting (and updates some things that fell out from testing it.)

I have focused on tests for the changes to iris/coords.py so far but will add tests for the changes to iris/plot.py once I have them (and once we have decided on what to do about discontiguity checking)

lbdreyer commented on lib/iris/tests/unit/plot/test_2d_coords.py:
```diff
@@ -1,157 +0,0 @@
-# (C) British Crown Copyright 2018, Met Office
```
I have removed this file as it had mixed unit tests for lots of different functions.
Unit tests should be separated so that each file tests a specific function/class (e.g. a single file for testing iris.plot._get_plot_defn)

I have moved the tests in this file to the appropriate files for *almost* all of them. Have not yet done it for the plot.py functions but will also push that up to this PR once that is done

corinnebosley replied:
Cool, it was on my list of things to do.  In my own PR I have shuffled some tests around and added some more for the new utils, but these were always intended to be cleaned up anyway, it's just a bit easier to run tests when they're in the same place.  

I guess I can pull in your changes and then move the rest of the tests round accordingly, because none of them are in the right place.

lbdreyer commented:
@pp-mo @corinnebosley 
Could someone look at this?

I have changed:
* Docs strings for a couple functions
* added support to cubes that may be of shape (2,1) or (1,2) or (1) to the discontiguity check
* changes `contiguous bounds` to check that the bounds are contiguous (before it was assumed)*
* Added a check in plot.py that the coordinates have the same number of dimensions (to stop someone doing a 1D coord against a 2D coord
* Removed the ability to modify the absolute tolerance when calling pcolormesh (instead use default)
* Added unit tests for:
  * coord methods`contiguous_bounds`, `is_contiguous` and `sanity_check_bounds`
  * `coords._discontiguity_in_2d_bounds`
  * `coords._get_2d_coord_bound_grid`
  * `plot.get_plot_defn`
  * `plot.get_plot_defn_custom_coords_picked`
* Added integration test for pcolormesh - will also need to add the expected graphic for this 

Note I have not written any tests for _check_contiguity_and_bounds or modified how we call that (other than to remove the abs_tol thing) but it does need fixing as currently it isn't right (we're calling it too many times) but this may be covered by what @corinnebosley is doing so I thought I'd leave it.

* Doing this has meant that when checking for discontiguities we are using the default absolute tolerance value which is 1e-8 which may be too small (it is too small for our ORCA2 test data).

pp-mo commented:
> it is too small for our ORCA2 test data

I'm worried about the practical usability of this.
If I understand right, between this and the bounds-check-routine that @corinnebosley is working on, this means that :
  * some ORCA data will generate warnings when plotted, because it has quite large discontiguities at certain points
  * the new contig-check routine will let you check the bounds with adjustable tolerance
  * but the only provided way to *avoid* the plot warning is to mask out the problem points. Which in fact you don't necessarily want to do.

So, I think it is necessary for the plot routine to either have a much more generous default tolerance, or a way to modify it.

IMHO the only point of warnings is to encourage people to take action to avoid them, so there should always _be_ a way to avoid it, and ideally the warning tells you how :  If there isn't a way out, then you are just giving more encouragement to ignore them !

corinnebosley commented on lib/iris/coords.py:
```diff
@@ -1043,9 +1066,9 @@ def _sanity_check_contiguous(self):
-                             '2 dimensions.'.format(self.name()))
+            raise ValueError('Invalid operation for {!r}. Not supported for '
+                             'bounds with more than 2 dimensions.'.format
+                             (self.name()))
```
I'm not sure about this error message.  It seems slightly more confusing than it was previously.

lbdreyer replied:
I think I changed this because I wanted the method `_sanity_check_bounds` to be able to be used for other things, but the previous error message.
Saying that, I realise I have made a lot of breaking changes in this PR, the above being one of them so I am going to undo that and put the error message back

corinnebosley commented on lib/iris/coords.py:
```diff
@@ -1064,28 +1087,25 @@ def is_contiguous(self, rtol=1e-05, atol=1e-08):
         return contiguous
 
     def contiguous_bounds(self):
         """
```
Docstring is better (I like things which are concise) but I'm still not taken with the name.  'contiguous_bounds' is just a bit vague.

lbdreyer replied:
Unfortunately we don't have much say with changing the name of `contiguous_bounds`. It was already part of the API so changing it would be a breaking change.

corinnebosley commented on lib/iris/coords.py:
```diff
@@ -1097,7 +1117,10 @@ def contiguous_bounds(self):
+            if not self.is_contiguous():
+                raise ValueError('Invalid operation. Bounds of coord {!r} '
+                                 'are not contiguous.'.format(self.name()))
+
```
Nice catch.  Much better.

lbdreyer replied:
Unfortunately this is a breaking change so I am going to have to put this back

corinnebosley commented on lib/iris/plot.py:
```diff
@@ -102,10 +102,20 @@ def get_span(coord):
+        if not (len(spans[0]) == len(spans[1])):
+            raise ValueError('The coordinate {!r} has {} dimensions and the '
+                             'coordinate {!r} has {} dimensions. Cell-based '
+                             'plotting is only supported for coordinates with '
```
I'd like to discuss this with you, because I am a little unsure of what cell-based plotting means. Also whether it is clear to users what cell-based plotting is.  And whether it matters, I suppose.

corinnebosley commented on lib/iris/plot.py:
```diff
@@ -273,10 +283,11 @@ def _invert_yaxis(v_coord, axes=None):
-    Check that the discontinuous bounds occur where the data is masked.
+    Checks that any discontiguity in the bounds of the coordinate only occur
+    where the data is masked.
 
```
This is neater and more accurate, but I have made some changes to the name and docstring for this function as well.  Probably worth agreeing on changes.

corinnebosley commented on lib/iris/tests/unit/coords/test_Coord.py:
```diff
@@ -355,6 +356,112 @@ def test_different_array_attrs_incompatible(self):
+    def test_1d_coord_contiguous_atol(self):
+        coord = iris.coords.DimCoord(self.points/1000, bounds=self.bounds/1000)
+        result = coord.is_contiguous(atol=1e-2)
+        self.assertTrue(result)
```
Might need you to explain this one.

corinnebosley commented on lib/iris/tests/unit/coords/test_Coord.py:
```diff
@@ -355,6 +356,112 @@ def test_different_array_attrs_incompatible(self):
+    def test_coord_1d_2_bounds(self):
+        coord = iris.coords.DimCoord([0, 1], standard_name='latitude',
+                                     bounds=np.array([[0, 1], [1, 2]]))
+        coord._sanity_check_bounds()
```
This is just to check that no error is raised is it?

lbdreyer replied:
Yep I'll add a comment explaining that

corinnebosley commented on lib/iris/tests/unit/coords/test_Coord.py:
```diff
@@ -355,6 +356,112 @@ def test_different_array_attrs_incompatible(self):
+                                     bounds=np.array([[0], [1]]))
+        emsg = "Invalid operation for 'latitude', with 1 bounds. Contiguous " \
+               "bounds are only defined for 1D coordinates with 2 bounds."
+        with self.assertRaisesRegexp(ValueError, emsg):
```
Shouldn't this raise an error about 2d coordinates needing 4 bounds, rather than 1d coordinates needing 2 bounds?

corinnebosley commented on lib/iris/tests/unit/coords/test__discontiguity_in_2d_bounds.py:
```diff
@@ -0,0 +1,143 @@
+        self.assertTrue(not diffs_along_x.any())
+        self.assertTrue(not diffs_along_y.any())
+
+    def test_discontiguous_along_x(self):
```
A short comment about what you are doing to the bounds array here might be useful.

corinnebosley commented on lib/iris/tests/unit/coords/test__discontiguity_in_2d_bounds.py:
```diff
@@ -0,0 +1,143 @@
+    def test_contiguous_abs_tol(self):
+        bds = self.discontiguous_bds / 1000
+        x_exp_result = np.array([1, 0]).reshape(2, 1) / 1000
+        y_exp_result = np.array([1, 0]).reshape(1, 2) / 1000
```
Don't really understand what's happening here.

lbdreyer commented:
**Summary of changes**

Discontiguity checking:
* Moved `_discontiguity_in_2d_bounds` function to a method on the coord so it could easily call other coord methods
* Rewrote `_discontiguity_in_2d_bounds` -> `_discontiguity_in_bounds` so that it supports 1D and 2D coords, and calls np.allclose
* `coord.is_contiguous` now calls `coord._discontiguity_in_bounds`. 1D coordinates are still handled as before (so this doesn't introduce a breaking change) but the handling has now been moved into  `_discontiguity_in_bounds` 

2d coord plotting:
* Introduced `contiguity_tolerance` kwarg to pcolor and pcolormesh
  * If it **is set**, the bounds of both 1D and 2D coords are checked for discontiguities using an absolute toelrance as set
  * If it **isn't set**, 1D coords are not checked, but 2D corods are checked with the defaults
* If discontiguities are found, an error is raised

Missing from this PR:
* An integration test but that will be easier to include in a separate PR

lbdreyer commented:
@pp-mo The tests are now passing, do you mind having a look?

pp-mo commented on lib/iris/plot.py:
```diff
@@ -271,54 +271,78 @@ def _invert_yaxis(v_coord, axes=None):
+            Absolute tolerance when checking the contiguity. Defaults to None.
+            If an absolute tolerance is not set, 1D coords are not checked (so
+            as to not introduce a breaking change without a major release) but
+            2D coords are checked with a relative tolerance.
```
> but 2D coords are checked with a relative tolerance.

This is not quite clear.  Maybe something like 
"but 2D coords are always checked, by calling `coord._discontiguity_in_bounds` with its default tolerance"
(If I understood it right)

lbdreyer commented on lib/iris/coords.py:
```diff
@@ -143,81 +143,43 @@ def __new__(cls, name_or_coord, minimum, maximum,
     Args:
-        bounds: array of shape (X,Y,4)
+    * bounds: (array)
+        Coordinate bounds array of shape (X,Y,4)
```
Ahh yeah sorry about that! I always get this the wrong way round!

lbdreyer commented on lib/iris/coords.py:
```diff
@@ -1064,40 +1090,33 @@ def is_contiguous(self, rtol=1e-05, atol=1e-08):
 
+        .. note::
+
+            Assumes bounds are contiguous
```
But if the 1d coord was already bounded it would assume the bounds are contiguous. I will clarify this point

lbdreyer commented on lib/iris/tests/unit/coords/test_Coord.py:
```diff
@@ -355,6 +356,279 @@ def test_different_array_attrs_incompatible(self):
+        self.assertArrayEqual(diffs_along_y, expected_diffs.reshape(0, 1))
+
+    def test_2d_one_cell_along_x(self):
+        # Test a 2D coord with a single cell along the x axis, where the coord
```
yep! got confused about (y,x) ordering again

lbdreyer commented on lib/iris/coords.py:
```diff
@@ -1064,40 +1090,33 @@ def is_contiguous(self, rtol=1e-05, atol=1e-08):
 
-        Assumes input is contiguous.
-
         .. note::
```
> Possibly worth stating that this fails for coords which are neither 1D nor 2D.

Will do!

> Or possibly make it return 'None', and say so.

Not keen on this. Is this not a breaking change.

lbdreyer commented on lib/iris/tests/unit/coords/test_Coord.py:
```diff
@@ -355,6 +356,279 @@ def test_different_array_attrs_incompatible(self):
+                                     bounds=[[1, 3], [4, 5], [5, 7]])
+        expected = np.array([1, 4, 5, 7])
+        result = coord.contiguous_bounds()
+        self.assertArrayEqual(result, expected)
```
I didn't include tests for 2d here is because `is_contiguous` is effectively a wrapper for calling `_discontiguity_in_bounds` and has things were already being tested in `Test__discontiguity_in_bounds` it seemed repetitive to also have them here.

I added the 1d cases to show that the behaviour is as before (before 1d was handled in `_discontiguity_in_bounds`.

I will add a couple of simple 2d examples and hopefully that will be enough testing.

lbdreyer commented on lib/iris/tests/unit/coords/test_Coord.py:
```diff
@@ -355,6 +356,279 @@ def test_different_array_attrs_incompatible(self):
+        coord = iris.coords.DimCoord(self.points,
+                                     bounds=self.discontiguous_bounds)
+        result = coord.is_contiguous(atol=1e1)
+        self.assertTrue(result)
```
By: 
> check that it uses '_discontinuity_in_bounds'

Do you mean to also remove the 1d cases and just have the test for non bounded coords and a test that it calls  `_discontinuity_in_bounds`?

lbdreyer commented on lib/iris/plot.py:
```diff
@@ -390,8 +408,8 @@ def _draw_2d_from_bounds(draw_method_name, cube, *args, **kwargs):
-        # If the data is tranposed, 2D coordinates will also need to be
-        # tranposed.
+        # If the data is transposed, 2D coordinates will also need to be
+        # transposed.
```
> Are we now allowing a mix of 1D of and 2D coordinates?

Sort of. We are allowing `get_plot_defn` to pick up a mix of 1D of and 2D coordinates. `_map_common` also supports a mix, but this part of the code does not. I can add a check that the coords are the same dimensionality here if you like.

lbdreyer commented:
@pp-mo  I believe I have addressed all your review comments in the most recent commit.

pp-mo commented:
> @lbdreyer pp-mo I believe I have addressed all your review comments in the most recent commit.

Yes, I'm happy that we are all done here now.
Result :+1: 

I'll merge, and hopefully all will check out, since #3150 is now also merged.
